service/internal/mapresolver/mapresolver.go

```diff
@@ -0,0 +1,211 @@
+
+	sync.Mutex
+	closers []config.CloseFunc
+	watcher chan error
+}
+
+// NewMapResolver returns a new MapResolver that resolves configuration from multiple locations.
+//
+// To resolve a configuration the following steps will happen:
+//   1. Retrieves individual configurations from all given "locations", and merge them in the retrieve order.
+//   2. Once the config.Map is merged, apply the converters in the given order.
+//
+// After the configuration was resolved the `MapResolver` can be used as a single point to watch for updates in
+// the configuration data retrieved via the config providers used to process the "initial" configuration and to generate
+// the "effective" one. The typical usage is the following:
+//
+//		mapResolver.Resolve(ctx)
+//		mapResolver.Watch() // wait for an event.
+//		mapResolver.Resolve(ctx)
+//		mapResolver.Watch() // wait for an event.
+//		// repeat Resolve/Watch cycle until it is time to shut down the Collector process.
+//		mapResolver.Shutdown(ctx)
+func NewMapResolver(set *Settings) (*MapResolver, error) {
+	if len(set.Locations) == 0 {
+		return nil, fmt.Errorf("cannot create ConfigProvider: no Locations")
+	}
+
+	if len(set.MapProviders) == 0 {
+		return nil, fmt.Errorf("cannot create ConfigProvider: no MapProviders")
+	}
+
+	// Safe copy, ensures the slices and maps cannot be changed from the caller.
+	locationsCopy := make([]string, len(set.Locations))
+	copy(locationsCopy, set.Locations)
+	mapProvidersCopy := make(map[string]config.MapProvider, len(set.MapProviders))
+	for k, v := range set.MapProviders {
+		mapProvidersCopy[k] = v
+	}
+	mapConvertersCopy := make([]config.MapConverterFunc, len(set.MapConverters))
+	copy(mapConvertersCopy, set.MapConverters)
+
+	return &MapResolver{
+		locations:           locationsCopy,
+		configMapProviders:  mapProvidersCopy,
+		configMapConverters: mapConvertersCopy,
+		watcher:             make(chan error, 1),
+	}, nil
+}
+
+// Resolve returns the configuration as a config.Map, or error otherwise.
+//
+// Should never be called concurrently with itself, Watch or Shutdown.
+func (mr *MapResolver) Resolve(ctx context.Context) (*config.Map, error) {
+	// First check if already an active watching, close that if any.
+	if err := mr.closeIfNeeded(ctx); err != nil {
+		return nil, fmt.Errorf("cannot close previous watch: %w", err)
+	}
+
+	// Retrieves individual configurations from all locations in the given order, and merge them in retMap.
+	retMap := config.NewMap()
+	for _, location := range mr.locations {
+		// For backwards compatibility:
+		// - empty url scheme means "file".
+		// - "^[A-z]:" also means "file"
+		scheme := "file"
+		if idx := strings.Index(location, ":"); idx != -1 && !driverLetterRegexp.MatchString(location) {
+			scheme = location[:idx]
+		} else {
+			location = scheme + ":" + location
+		}
+		p, ok := mr.configMapProviders[scheme]
+		if !ok {
+			return nil, fmt.Errorf("scheme %v is not supported for location %v", scheme, location)
```

Comment:
Perhaps `no providers found for the scheme %q` would better convey the situation?

Reply:
(although looks like this comes from an existing code)